### PR TITLE
cnf network: Add Rdma Metrics Test Cases

### PR DESCRIPTION
**Pull request by ajaggapa**

### CODE REVIEW

File: tests/cnf/core/network/sriov/tests/rdmametrics.go [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+					[]string{"bash", "-c", fmt.Sprintf("ls /sys/bus/pci/devices/%s/infiniband/%s/ports/1/hw_counters",
+						pciAddress, rdmaDevice)})
+				Expect(err).To(HaveOccurred(), "Expected command to fail as the path is not present on Pod")
+				Expect(podOutput.String()).To(ContainSubstring("No such file or directory"),
+					fmt.Sprintf("Expected error 'No such file or directory' in the output %s", podOutput.String()))
+
+				By("Verify Rdma Metrics should be present on Host")
+				nodeOutput, err := cluster.ExecCmdWithStdout(APIClient,
+					fmt.Sprintf("ls /sys/bus/pci/devices/%s/infiniband/%s/ports/1/hw_counters", pciAddress, rdmaDevice),
+					metav1.ListOptions{LabelSelector: fmt.Sprintf("kubernetes.io/hostname=%s", workerNodeList[0].Object.Name)})
```
Comment (evgenLevin) [inline tests/cnf/core/network/sriov/tests/rdmametrics.go:178]: please replace workerNodeList[0].Object.Name by  testPod.Object.Spec.NodeName

File: tests/cnf/core/network/sriov/tests/rdmametrics.go [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+				},
+			},
+		},
+	}
+
+	finalIgnitionConfig, err := json.Marshal(ignitionConfig)
+	Expect(err).ToNot(HaveOccurred(), "Failed to serialize ignition config")
+
+	createdMC, err := mco.NewMCBuilder(APIClient, "02-rdma-netns-exclusive-mode").
+		WithLabel("machineconfiguration.openshift.io/role", "workercnf").
```
Comment (evgenLevin) [inline tests/cnf/core/network/sriov/tests/rdmametrics.go:235]: please use CnfMcpLabel
https://github.com/openshift-kni/eco-gotests/blob/main/tests/cnf/core/network/internal/netconfig/config.go#L29

File: tests/cnf/core/network/sriov/tests/rdmametrics.go [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+		WithRawConfig(finalIgnitionConfig).
+		Create()
+	Expect(err).ToNot(HaveOccurred(), "Failed to create RDMA machine config")
+
+	return createdMC
+}
+
+func getInterfacePci(tPod *pod.Builder, podInterface string) (string, string, error) {
```
Comment (evgenLevin) [inline tests/cnf/core/network/sriov/tests/rdmametrics.go:243]: can we reuse getPCIAddressListFromSrIovNetworkName for rootless.go?

File: tests/cnf/core/network/sriov/tests/rdmametrics.go [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+		}
+	}
+
+	tPod, err := pod.NewBuilder(APIClient, "testpod", tsparams.TestNamespaceName, NetConfig.CnfNetTestContainer).
+		WithSecondaryNetwork(netAnnotation).
+		WithPrivilegedFlag().
+		CreateAndWaitUntilRunning(2 * time.Minute)
+	Expect(err).ToNot(HaveOccurred(), "Failed to create Pod %s")
+
+	tPod.Exists()
```
Comment (evgenLevin) [inline tests/cnf/core/network/sriov/tests/rdmametrics.go:340]: ?

File: tests/cnf/core/network/sriov/tests/rdmametrics.go [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+		testNetBuilder.WithMetaPluginRdma()
+	}
+
+	testNetwork, err := testNetBuilder.Create()
+	Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("Failed to create Sriov Network %s", netName))
+
+	return testNetwork
+}
+
+func defineAndCreatePod(netName1, netName2 string) *pod.Builder {
```
Comment (evgenLevin) [inline tests/cnf/core/network/sriov/tests/rdmametrics.go:309]: just a suggestion


```
func defineAndCreatePod(netName1, netName2 ...string) *pod.Builder {

netAnnotations := []*types.NetworkSelectionElement{
    {
        Name:       netName1,
        MacRequest: tsparams.ServerMacAddress,
        IPRequest:  []string{tsparams.ServerIPv4IPAddress},
    },
}

if len(netName2) != 0 {
    netAnnotations = append(netAnnotations, &types.NetworkSelectionElement{
        Name:       netName2[0],
        MacRequest: tsparams.ClientMacAddress,
        IPRequest:  []string{tsparams.ClientIPv4IPAddress},
    })
}

```

File: tests/cnf/core/network/sriov/tests/rdmametrics.go [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+
+	return tPod
+}
+
+func verifyRdmaMetrics(inputPod *pod.Builder, iName string) {
+	By("Fetch PCI Address and Rdma device from Pod Annotations")
+
+	pciAddress, rdmaDevice, err := getInterfacePci(inputPod, iName)
+	Expect(err).ToNot(HaveOccurred(),
+		fmt.Sprintf("Could not get PCI Address and/or Rdma device from Pod Annotations with Error: %s", err))
```
Comment (evgenLevin) [inline tests/cnf/core/network/sriov/tests/rdmametrics.go:350]: no need err in the description

**Review by evgenLevin — COMMENTED** [review]
overall looks good. Just several comments

File: tests/cnf/core/network/sriov/internal/tsparams/consts.go [hunk 7]
```diff
@@ -15,4 +15,6 @@ const (
 	LabelExposeMTUTestCases = "exposemtu"
 	// LabelSriovMetricsTestCases represents Sriov Metrics Exporter label that can be used for test cases selection.
 	LabelSriovMetricsTestCases = "sriovmetrics"
+	// LabelRdmaMetricsTestCases represents Sriov Metrics Exporter label that can be used for test cases selection.
```
[the remark this answers is not in the document]
Comment (ajaggapa) [reply tests/cnf/core/network/sriov/internal/tsparams/consts.go:18]: done

File: tests/cnf/core/network/sriov/tests/rdmametrics.go [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+var _ = Describe("rdmaMetrics", Ordered, Label(tsparams.LabelRdmaMetricsTestCases),
+	ContinueOnFailure, func() {
+		var (
+			workerNodeList           []*nodes.Builder
+			sriovInterfacesUnderTest []string
+			sriovDeviceID            string
```
[the remark this answers is not in the document]
Comment (ajaggapa) [reply tests/cnf/core/network/sriov/tests/rdmametrics.go:33]: done

File: tests/cnf/core/network/sriov/tests/rdmametrics.go [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+			rdmaMachineConfig        *mco.MCBuilder
+		)
+
+		BeforeAll(func() {
+			By("Verifying if Sriov Metrics Exporter tests can be executed on given cluster")
```
[the remark this answers is not in the document]
Comment (ajaggapa) [reply tests/cnf/core/network/sriov/tests/rdmametrics.go:38]: done

File: tests/cnf/core/network/sriov/tests/rdmametrics.go [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+			err := netenv.DoesClusterHasEnoughNodes(APIClient, NetConfig, 1, 1)
+			Expect(err).ToNot(HaveOccurred(),
+				"Cluster doesn't support Sriov Metrics Exporter test cases as it doesn't have enough nodes")
```
[the remark this answers is not in the document]
Comment (ajaggapa) [reply tests/cnf/core/network/sriov/tests/rdmametrics.go:41]: done

File: tests/cnf/core/network/sriov/tests/rdmametrics.go [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+				By("Define and Create Test Resources")
+
+				tPol := defineAndCreateNodePolicy("rdmapolicy1", "sriovpf1", sriovInterfacesUnderTest[0], 1, 0)
+				tNet := defineAndCreateSriovNetworkWithRdma("sriovnet1", tPol.Object.Spec.ResourceName, false)
+				testPod := defineAndCreatePod(tNet.Object.Name, "")
+
+				By("Fetch PCI Address and Rdma device from Pod Annotations")
+				pciAddress, rdmaDevice, err := getInterfacePci(testPod, "net1")
+				Expect(err).ToNot(HaveOccurred(),
+					fmt.Sprintf("Could not get PCI Address and/or Rdma device from Pod Annotations with Error: %s", err))
```
[the remark this answers is not in the document]
Comment (ajaggapa) [reply tests/cnf/core/network/sriov/tests/rdmametrics.go:165]: done

File: tests/cnf/core/network/sriov/tests/rdmametrics.go (hunk 2 above)
Comment (ajaggapa) [reply tests/cnf/core/network/sriov/tests/rdmametrics.go:235]: done

File: tests/cnf/core/network/sriov/tests/rdmametrics.go (hunk 3 above)
Comment (ajaggapa) [reply tests/cnf/core/network/sriov/tests/rdmametrics.go:243]: We need **_rdma-device_** along with **_pci-address_**, but rootless function gives only pci addresses.

File: tests/cnf/core/network/sriov/tests/rdmametrics.go (hunk 4 above)
Comment (ajaggapa) [reply tests/cnf/core/network/sriov/tests/rdmametrics.go:340]: `CreateAndWaitUntilRunning()` returns the pod builder which is still in Pending status. So it does not have network-status annotations set. `Exists()` updates the builder with all fields, so we can see the annotations.

Ideally this should be part of `CreateAndWaitUntilRunning` method.

File: tests/cnf/core/network/sriov/tests/rdmametrics.go (hunk 6 above)
Comment (ajaggapa) [reply tests/cnf/core/network/sriov/tests/rdmametrics.go:350]: done

**Review by evgenLevin — DISMISSED** [review]
lgtm

File: tests/cnf/core/network/sriov/tests/rdmametrics.go [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+	"github.com/openshift-kni/eco-goinfra/pkg/sriov"
+	"github.com/openshift-kni/eco-gotests/tests/cnf/core/network/internal/netenv"
+	. "github.com/openshift-kni/eco-gotests/tests/cnf/core/network/internal/netinittools"
+	"github.com/openshift-kni/eco-gotests/tests/cnf/core/network/internal/netparam"
+	"github.com/openshift-kni/eco-gotests/tests/cnf/core/network/sriov/internal/sriovenv"
+	"github.com/openshift-kni/eco-gotests/tests/cnf/core/network/sriov/internal/tsparams"
+	"github.com/openshift-kni/eco-gotests/tests/internal/cluster"
+	"gopkg.in/k8snetworkplumbingwg/multus-cni.v4/pkg/types"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
```
Comment (SchSeba) [inline tests/cnf/core/network/sriov/tests/rdmametrics.go:25]:
```suggestion
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/labels"
	
	ignition "github.com/coreos/ignition/v2/config/v3_1/types"
	"gopkg.in/k8snetworkplumbingwg/multus-cni.v4/pkg/types"
	
	"github.com/openshift-kni/eco-goinfra/pkg/mco"
	"github.com/openshift-kni/eco-goinfra/pkg/namespace"
	"github.com/openshift-kni/eco-goinfra/pkg/nodes"
	"github.com/openshift-kni/eco-goinfra/pkg/pod"
	"github.com/openshift-kni/eco-goinfra/pkg/reportxml"
	"github.com/openshift-kni/eco-goinfra/pkg/sriov"
	"github.com/openshift-kni/eco-gotests/tests/cnf/core/network/internal/netenv"
	. "github.com/openshift-kni/eco-gotests/tests/cnf/core/network/internal/netinittools"
	"github.com/openshift-kni/eco-gotests/tests/cnf/core/network/internal/netparam"
	"github.com/openshift-kni/eco-gotests/tests/cnf/core/network/sriov/internal/sriovenv"
	"github.com/openshift-kni/eco-gotests/tests/cnf/core/network/sriov/internal/tsparams"
	"github.com/openshift-kni/eco-gotests/tests/internal/cluster"
	

```

File: tests/cnf/core/network/sriov/tests/rdmametrics.go [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+
+			sriovInterfacesUnderTest, err = NetConfig.GetSriovInterfaces(2)
+			Expect(err).ToNot(HaveOccurred(), "Failed to retrieve SR-IOV interfaces for testing")
+
+			By("Fetching SR-IOV Device ID for interface under test")
+			sriovDeviceID := discoverInterfaceUnderTestDeviceID(sriovInterfacesUnderTest[0], workerNodeList[0].Definition.Name)
+			Expect(sriovDeviceID).ToNot(BeEmpty(), "Expected sriovDeviceID not to be empty")
+
+			By("Skipping test cases if the Sriov device is not of Mellanox")
+			if !(sriovDeviceID == netparam.MlxDeviceID || sriovDeviceID == netparam.MlxBFDeviceID) {
```
Comment (SchSeba) [inline tests/cnf/core/network/sriov/tests/rdmametrics.go:58]: can can check the vendorID as any of the mlx devices support RDMA

**Review by SchSeba — COMMENTED** [review]
lgtm just two small comments

**Comment by gkopels** [comment]
lgtm

**Review by gkopels — APPROVED** [review]

**Review by evgenLevin — APPROVED** [review]

**Pull request merged**